### DISCUSSION TITLE
Ethernet - pins being pulled down after Ethernet.begin()

### DISCUSSION THREAD

**Discussion by xRadx** [stm32duino/Arduino_Core_STM32#2262] [Libraries] [answered]
Hi,

have anyone faced such situation that following pins are begin pulled down:
PB11, PB12, PB5, PC2, PE2, PG14, PG8 after calling Ethernet.begin() ?

Firstly I init Serial, Serial4, Serial5, then most pins avaliable on the board as input_pullup. 
Then I read them and all are high. Next I call Ethernet.begin() and read all the pins again and the above ones are low (the rest stays as high)...

I use Nucleo F746GZ.

**Comment by fpistm** [comment, marked as answer]
Hi,
by defauolt Ethernet pins are set to PU:
https://github.com/stm32duino/Arduino_Core_STM32/blob/50629bd31f2edc873363f0be13ce012c6974d9ff/variants/STM32F7xx/F745Z(E-G)T_F746Z(E-G)(T-Y)_F750Z8T_F756ZG(T-Y)/PeripheralPins.c#L391-L417

but probably the HAL set it to down.
Note that you have to redefine this weak array with the pins used for ETH.
See https://github.com/stm32duino/Arduino_Core_STM32/wiki/Custom-definitions#custom-pinmap-array

**Reply by xRadx** [reply]
I made a file "build_opt.h" (in the sketch folder) with lines:
const PinMap PinMap_Ethernet[] = { 
   {PA_1,      ETH, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF11_ETH)},
   {PA_2,      ETH, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF11_ETH)}, 
   {PA_7,      ETH, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF11_ETH)},
   {PB_13,     ETH, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF11_ETH)},
   {PC_1,      ETH, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF11_ETH)},
   {PC_4,      ETH, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF11_ETH)},
   {PC_5,      ETH, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF11_ETH)},
   {PG_11,     ETH, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF11_ETH)},
   {PG_13,     ETH, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF11_ETH)},
   {NC,        NP,  0}
}
(these pins used are as generated by CubeMX and it works in a CubdeIDE) and included (#include "build_opt.h") in the project.ino file.

However it does not compile and gives errors:
arm-none-eabi-g++: warning: const: linker input file unused because linking not done
arm-none-eabi-g++: error: const: linker input file not found: No such file or directory
arm-none-eabi-g++: warning: PinMap: linker input file unused because linking not done
arm-none-eabi-g++: error: PinMap: linker input file not found: No such file or directory
arm-none-eabi-g++: warning: PinMap_Ethernet[]: linker input file unused because linking not done
arm-none-eabi-g++: error: PinMap_Ethernet[]: linker input file not found: No such file or directory
...
arm-none-eabi-g++: warning: {PA_1,: linker input file unused because linking not done
arm-none-eabi-g++: error: {PA_1,: linker input file not found: No such file or directory
arm-none-eabi-g++: warning: ETH,: linker input file unused because linking not done
arm-none-eabi-g++: error: ETH,: linker input file not found: No such file or directory
...

I was trying to find a help/tips on internet/youtube, but nothing found...

**Reply by fpistm** [reply]
No build_opt.h is not a C file. Put your new array on top of your sketch.

**Reply by xRadx** [reply]
Now it works. Thanks!
